### ISSUE TITLE
Finetuning AIS decoder (continued)

### ISSUE THREAD

**Issue by chuyaowang** [computational-cell-analytics/micro-sam#1211]
Hi, continuing from this issue #1205, here is an example of our data:

<img width="2596" height="1476" alt="Image" src="https://github.com/user-attachments/assets/bec38fc1-922d-4353-9883-3c133b9e348f" />

The size is 2304 x 2304 pixels, imaged with DIC microscopy. Here I am using tile size of (512,512) and halo (64,64). From the boundary distance, the hypha cells can be resolved, but the watershedding post processing cannot handle overlaps. The foreground probability and center distance map are also not 100% accurate. I want to fine-tune the decoder to improve the 3 outputs then use other algorithms to join the oversegmented pieces of the hypha cells. Right now I have about 20 images like this, I will also tile them and apply augmentations. From your experience, what's the minimum number of images that will suffice for finetuning, as labeling these images with instance masks are quite labor intensive? Also, do you think it's better to label a cell split by a crossing point with one mask (treat as same cell) or two masks (treat as two cells)?

**Comment by anwai98** [comment]
Hi @chuyaowang,

Thanks for sharing the details.

To quickly step back and talk about the preliminary details:

> The size is 2304 x 2304 pixels, imaged with DIC microscopy. Here I am using tile size of (512,512) and halo (64,64). 

I see. On a higher level, I'd give `(384, 384)` a quick shot and see if the distance predictions are better resolved (I doubt it, but assuming you are using the `*_lm` generalist models, this would be a good starting point to be consistent with pretrained models' field of view)

> From the boundary distance, the hypha cells can be resolved, but the watershedding post processing cannot handle overlaps.

Yeah, we've noticed that when such filamental structures overlap a lot, it's quite hard to resolve them. One quick thing I'd recommend would be to see if finetuning helps (I'll answer your questions on it below)

> I want to fine-tune the decoder to improve the 3 outputs then use other algorithms to join the oversegmented pieces of the hypha cells. Right now I have about 20 images like this, I will also tile them and apply augmentations. From your experience, what's the minimum number of images that will suffice for finetuning, as labeling these images with instance masks are quite labor intensive?

Finetuning sounds like a right direction. To begin with, I'd say annotate like 5-10 images (use `micro-sam` for it, and try to disentangle intertwined cells as good as possible). And I'd recommend finetuning the entire model, i.e. both encoder and decoder on your data to get the most optimal performance. 

> Also, do you think it's better to label a cell split by a crossing point with one mask (treat as same cell) or two masks (treat as two cells)?

Speaking in biological context, I personally think you should annotate the cell splits as one, in case the split was due to another overlapping cell (since biologically speaking, it's still the same cell - so treating them as one cell is the way-to-go). It's a different story if `micro-sam` (or other methods) can resolve it or not.

Hope this helps. And let us know how this goes!

**Comment by constantinpape** [comment]
Just to add some thought from my end:


> I see. On a higher level, I'd give `(384, 384)` a quick shot and see if the distance predictions are better resolved (I doubt it, but assuming you are using the `*_lm` generalist models, this would be a good starting point to be consistent with pretrained models' field of view)

I wouldn't go smaller than what you're using so far (i.e. 512 x 512 with halo 64 x 64). It will be difficult to correctly predict the center position otherwise in too small patches. This implies training with a patch shape of 512 + 128 = 640.

> From the boundary distance, the hypha cells can be resolved, but the watershedding post processing cannot handle overlaps

Yes, that is an in-built limitation of our post-processing approach. You will not be able to improve this aspect with fine-tuning alone and will either need to add more post-processing to merge the filamentous cells (as you already wrote).

> Right now I have about 20 images like this, I will also tile them and apply augmentations. From your experience, what's the minimum number of images that will suffice for finetuning, as labeling these images with instance masks are quite labor intensive?

20 images should be more than enough. Note: You don't need to do tiling or augmentations yourself, our training pipeline already takes care of this. As @anwai98 wrote, 5-10 would probably be enough (but using the 20 you have doesn't hurt of course).

> Also, do you think it's better to label a cell split by a crossing point with one mask (treat as same cell) or two masks (treat as two cells)?

Label them with one mask. This is biologically correct and you would want this for correct evaluation or methods that can deal with the over-load. Our data loader will automatically split non-touching objects internally.

**Comment by chuyaowang** [comment]
Thank you for your feedbacks. I'm indeed using the vit_l_lm model, I forgot to mention. Good to know augmentation and tiling are already built in! I only have 6GB of vram, so I'll first finetune the decoder using the embeddings from the vit_l_lm encoder. Then finetune a smaller encoder and the decoder together and see which one works better.

**Comment by anwai98** [comment]
Hi @chuyaowang,

Sounds great. Let us know what your finds are. For now, I'll consider this answered and close the issue.

If you have any further questions, please do not hesitate to reach out to us! (feel free to open a new issue for further discussions)

**Comment by constantinpape** [comment]
> I'm indeed using the vit_l_lm model

FYI: We recommend using the `vit_b_lm` model. It's smaller, i.e. you will need less RAM, and as good as `vit_l_lm`.
